modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/ExecutionTargetProviderImpl.java

```diff
@@ -152,22 +162,45 @@ private CompletableFuture<TokenizedAssignments> primaryReplica(
         });
     }
 
-    private CompletableFuture<TokenizedAssignments> allReplicas(
-            ReplicationGroupId replicationGroupId,
+    private CompletableFuture<List<TokenizedAssignments>> allReplicas(
+            List<TablePartitionId> replicationGroupIds,
             HybridTimestamp operationTime
     ) {
-        CompletableFuture<TokenizedAssignments> f = placementDriver.getAssignments(
-                replicationGroupId,
+        CompletableFuture<List<TokenizedAssignments>> f = placementDriver.getAssignments(
+                replicationGroupIds,
                 operationTime
         );
 
         return f.thenCompose(assignments -> {
-            if (assignments == null) {
-                // assignments are not ready yet, let's fall back to primary replicas
-                return primaryReplica(replicationGroupId, operationTime);
+            // Collect missed assignments indexes if found.
+            IntList missedAssignments = new IntArrayList(0);
+
+            for (int i = 0; i < assignments.size(); i++) {
+                if (assignments.get(i) == null) {
+                    missedAssignments.add(i);
+                }
+            }
+
+            if (missedAssignments.isEmpty()) {
+                return completedFuture(assignments);
             }
 
-            return completedFuture(assignments);
+            // assignments are not ready yet, let's fall back to primary replicas
+            List<CompletableFuture<TokenizedAssignments>> primaryReplicaAssignment = new ArrayList<>(missedAssignments.size());
+
+            for (int i = 0; i < missedAssignments.size(); i++) {
+                primaryReplicaAssignment.add(primaryReplica(replicationGroupIds.get(missedAssignments.getInt(i)), operationTime));
+            }
+
+            CompletableFuture<Void> all = CompletableFuture.allOf(primaryReplicaAssignment.toArray(new CompletableFuture[0]));
+            return all.thenApply(ignore -> {
+                // Replace missed assignments with primary replicas.
+                for (int i = 0; i < missedAssignments.size(); i++) {
+                    assignments.set(missedAssignments.getInt(i), primaryReplicaAssignment.get(i).join());
```

Comment:
In `AssignmentsTracker` you used `Collectors.toList()` to collect assignments into a list. Javadoc of this collectors states, that `There are no guarantees on the type, mutability, serializability, or thread-safety of the {@code List} returned`. I think, it's pretty reasonable contract for `AssignmentsTracker`, but in order to make `ExecutionTargetProviderImpl`, it's better to explicitly make sure that list is mutable. Let's create new _mutable_ list in case we need to fallback to primary replica assignments for certain partitions